### DISCUSSION TITLE
How to get E0s for VASP users

### DISCUSSION THREAD

**Discussion by liubaoshuai1402** [ACEsuit/mace#715] [Q&A] [answered]
Dear developers,

  I try to get E0s, the atomic energy by calculating the single point energy of a single atom which is placed in a large cell.
  But the value is slightly different from the defaults when i set E0s: "foundation". I am trying fine tuning model of MP.
  For example, the default value of Zr atom is -2.28579303 eV, but my own calculated results is  -2.09089 eV.
  Therefore, I wonder to know
 (1) The way i calculated my own E0s is right?
 (2) the difference from the defaults and my own calculated results is acceptable?

**Comment by gabor1** [comment, marked as answer]
you may be using a different set of pseudopotentials, or one or more of your other settings is different from those that were used to compute the Materials Project data. The question is what you use your newly computed E0s for. Are you fine-tuning the model? If your VASP parameters are consistent within your new dataset (E0s and all), that's all that matters.

**Reply by liubaoshuai1402** [reply]
Yes, I am try Fine-tuning Foundation Models, these parameters are consistent except spin polarization.
I follow the user guide of MACE, and perform spin polarized calculations when calculating my own E0s.
But when I perform calculations of bulk phase, I don't open the spin polarization because i think my system don't have magnetism.
Is it OK?

**Reply by gabor1** [reply]
if you are _sure_ that a spin polarised calculation in your bulk would just return zero spin and therefore the same energies and forces as a non-spin-polarised calculation, then yes of course you can run those without spin-polarisation.

**Reply by liubaoshuai1402** [reply]
Thanks, it really help me a lot !

**Comment by nihil39** [comment]
> Dear developers,
> 
> I try to get E0s, the atomic energy by calculating the single point energy of a single atom which is placed in a large cell. But the value is slightly different from the defaults when i set E0s: "foundation". I am trying fine tuning model of MP. For example, the default value of Zr atom is -2.28579303 eV, but my own calculated results is -2.09089 eV. Therefore, I wonder to know (1) The way i calculated my own E0s is right? (2) the difference from the defaults and my own calculated results is acceptable?

Hi, 

just a curiosity: where do you get the data for your default value (-2.285... eV)?  I know that if you use PBE functional you should get a value close to zero, as shown in the Materials Project website. I am running the same calculations (isolated atom in a large cell) for Oxygen and Hydrogen and I got values near zero for PBE.

Thanks

**Reply by liubaoshuai1402** [reply]
If you try fine tuning model of MP with multihead training mode (added in MACE0.37 and after) and set Eos 'foundation', you should find these defaults in log file.

**Reply by liubaoshuai1402** [reply]
Besides, you said that you got values near zero for PBE, but how much are these values ? -0.8 or -0.01, I don's think former value is close to zero.

**Reply by nihil39** [reply]
RevPBE-D3 
Hydrogen: -0.11941284 eV
Oxygen: -0.43787111 eV
 
 PBE
 energy  without entropy=       -0.00008751 eV (Hydrogen)
 energy  without entropy=       -0.00457539 eV (Oxygen)

**Reply by bernstei** [reply]
From https://www.vasp.at/vasp-workshop/pseudoppdatabase.pdf 
"in VASP all energies are referred to the the reference state for which the potential
was generated!
this is in most cases not the real groundstate of the atom"

MACE MP0's E0s are the results of actual VASP calculations of the isolated atom with pymatgen MPStaticSet values, i.e. their choice of initial magmoms, spin polarized, with the same cutoff as the mptrj fitting data

**Reply by liubaoshuai1402** [reply]
> PBE energy without entropy= -0.00008751 eV (Hydrogen) energy without entropy= -0.00457539 eV (Oxygen)

Hi, these energie are really close to zero, but are different from my calculated values(Tomorrow i will show my results it too night in my country)

**Reply by nihil39** [reply]
Ok, thanks. Can you also attach your vasp input files?

**Reply by liubaoshuai1402** [reply]
> Ok, thanks. Can you also attach your vasp input files?
[INCAR.txt](https://github.com/user-attachments/files/18087935/INCAR.txt)
<img width="964" alt="energy" src="https://github.com/user-attachments/assets/564490df-1d1a-4405-a0db-0686749d8a51">
This figure shows the Eos of oxygen I have calculated.

**Reply by nihil39** [reply]
Thanks, maybe the difference is due to the fact that you are using Revised PBE for solids (PBEsol) functional (the same used in materials project dataset?) while we are using pure PBE (`GGA = PE`)  and RevPBE-D3 (`GGA=RE`). 

Can you also provide POSCAR, POTCAR and KPOINTS files? I will attach ours as soon as possible.

**Reply by liubaoshuai1402** [reply]
> Thanks, maybe the difference is due to the fact that you are using Revised PBE for solids (PBEsol) functional (the same used in materials project dataset?) while we are using pure PBE (`GGA = PE`) and RevPBE-D3 (`GGA=RE`).
> 
> Can you also provide POSCAR, POTCAR and KPOINTS files? I will attach ours as soon as possible.


Here are related files. In fact, I don's think the choice of potential can lead such large difference !
[POSCAR.txt](https://github.com/user-attachments/files/18094236/POSCAR.txt)
[OUTCAR.txt](https://github.com/user-attachments/files/18094239/OUTCAR.txt)
[POTCAR.txt](https://github.com/user-attachments/files/18094241/POTCAR.txt)
[KPOINTS.txt](https://github.com/user-attachments/files/18094245/KPOINTS.txt)

**Reply by nihil39** [reply]
[vasp_input_file_oxygen_isolated.tar.gz](https://github.com/user-attachments/files/18094723/vasp_input_file_oxygen_isolated.tar.gz)
These are our files.

Maybe the difference is due to the different cell size (we used 15 Angstroms per side), 
number of k points (why did you use 27 if we are calculating the energy of an isolated atom?), 
the fact that we didn't use spin polarized calculations 
and the difference between the functional used in the POTCAR file and the one specified in the INCAR file (in our calculations it was pure PBE in both files)

**Reply by liubaoshuai1402** [reply]
> Maybe the difference is due to the different cell size (we used 15 Angstroms per side), number of k points (why did you use 27 if we are calculating the energy of an isolated atom?), the fact that we didn't use spin polarized calculations and the difference between the functional used in the POTCAR file and the one specified in the INCAR file (in our calculations it was pure PBE in both files)

I am not a VASP expert, for KPOINTS and cell size, they are chose to keep consistent to the parameters of my bulk phase.

**Comment by ftherrien** [comment]
@gabor1 could you explain briefly *why* it is so important that the E0s are consistent with the new dataset or even why it is important that they are precise? If mace learns a correction to these energies, isn't the only truly important thing that the E0s are consistent between training and inference?

**Reply by ilyes319** [reply]
In most MACE models we use (including the foundation models), we do not let the model learn a shift to the potential energy, and the only shift is in the E0s. This is made to ensure correct limits when dissociating atoms. Therefore, any shift needs to be compensated by the model itself and can limit its expressivity to fit the actual data. It is important that the E0s are as precise as possible, to not mess up the finetuning. When training from sratch, it might be less important, but still is.

**Reply by gabor1** [reply]
A key technical point is that since an isolated atom has no neighbours, no graph, there is a specific number that a mace model will return as the energy of an isolated atom. If you use the --model=MACE (the default) option, this number will be the E0 that you give either on the command line or as part of the input data if it contains a frame with an isolated atom. If you use the --model=ScaleShiftMACE, or the --E0=average with the MACE model, then the algorithm will make the best element-dependent shift that it find gives the lowest loss, and  depending on the composition of your dataset, it will be some kind of average binding energy. In typical condensed phase simulations, this will be many eVs lower than the isolated atom energy. This will result in very unphysical bond dissociation curves, low barriers for bond dissociation etc. 

If you give an imprecise E0 that is close to the real isolated atom energy, then the resulting potential energy surface will not be very unstable, you'll just get the wrong energy for an isolated atom. So if ever in your simulations or in your training set, you do get an isolated atom (even as part of a big cell but with nothing within cutoff), you'll get the wrong answer. 

The reason you need to give the ACTUAL spin-polarised isolated atom energy E0 when you fine-tune the foundation model is that the original foundation model used the actual isolated atom energies as E0s, so it made a model with millions of weights approximating the binding energy with that as reference. When you now fine-tune, if you demand that the reference is also changed, then you are asking for extra effort. It might work, but you might need more data, more training etc.

**Reply by gabor1** [reply]
BTW these are not really mace-specific ideas, they apply to most (all?) local models.

**Reply by ftherrien** [reply]
> If you give an imprecise E0 that is close to the real isolated atom energy, then the resulting potential energy surface will not be very unstable, you'll just get the wrong energy for an isolated atom. So if ever in your simulations or in your training set, you do get an isolated atom (even as part of a big cell but with nothing within cutoff), you'll get the wrong answer. 

So in inorganic crystal systems it would be less important to get precise E0s, because you would never get isolated atoms. I am understanding this correctly?

> The reason you need to give the ACTUAL spin-polarised isolated atom energy E0 when you fine-tune the foundation model is that the original foundation model used the actual isolated atom energies as E0s, so it made a model with millions of weights approximating the binding energy with that as reference. When you now fine-tune, if you demand that the reference is also changed, then you are asking for extra effort. It might work, but you might need more data, more training etc.

and

> Therefore, any shift needs to be compensated by the model itself and can limit its expressivity to fit the actual data. It is important that the E0s are as precise as possible, to not mess up the finetuning.

~~Isn't this also in some way an argument for *not* changing the E0s at all (using the "foundation" option)? The model now has to learn from a new reference *and* learn a new shift.~~

I think what you are saying is that if we can shift the reference (E0s) by an amount that is perfectly consistent with the new method, than the model has less "newness" to learn (i.e. most of the difference between the new method and the one used for the foundation model might be in the E0s). As pointed out by @nihil39 different choices of "isolated atoms" give pretty significantly different results, so in order for precise E0s to really help, they would need to be perfectly consistent with (1) the new method used *and* (2) with how you originally calculated them (i.e. your POSCARs) no?

**Reply by bernstei** [reply]
> Isn't this also in some way an argument for not changing the E0s at all (using the "foundation" option)? The model now has to learn from a new reference and learn a new shift.

This is definitely not true.  The only thing that's at all likely to be transferable to a new calculation is the binding energy.  For example, your fitting energies could be using CASTEP, which includes the core energy in the total energy and is off by hundreds of eV, despite having similar physics and similar binding energies.

>  so in order for precise E0s to really help, they would need to be perfectly consistent with (1) the new method used and (2) with how you originally calculated them (i.e. your POSCARs) no?

What do you mean by "originally calculated them"? Originally for the MACE universal pretraining data, or for your own (fine tuning) calculations?

**Reply by ftherrien** [reply]
Agreed, I see how that does not make sense in most cases


I mean the MACE original pretraining.
